### PR TITLE
Do not inappropriately register bind when the value is a closure

### PR DESCRIPTION
**Pull request by vlakoff**
Prerequisite for #5138.

The main fix is commit https://github.com/vlakoff/CodeIgniter4/commit/223ca28400a95594490052b032e1a921237138fb.

### CODE REVIEW

**Comment by kenjis** [comment]
Unit testing tests the object behavior, not in the code, not the specific method.

**Comment by kenjis** [comment]
According to line coverage, there are two lines 682,683 in `whereHaving()` that are not covered. 
Screenshot 2021-10-30 11 40 25

**Comment by kenjis** [comment]
The branch coverage for `whereHaving()` is 84.38%.
Not bad.

```
$ ./phpunit --coverage-html=build/coverage --path-coverage tests/system/Database/Builder/ -d memory_limit=1024m
```

Screenshot 2021-10-30 13 15 55

**Comment by kenjis** [comment]
@vlakoff Adding this test gets 100% line coverage.

```php
    public function testWhereAssociateArrayKeyHasEqualValueIsNull()
    {
        $builder = $this->db->table('user');

        $where = [
            'id <'   => 100,
            'col1 =' => null,
        ];
        $builder->where($where);

        $expectedSQL = 'SELECT * FROM "user" WHERE "id" < 100 AND "col1" IS NULL';
        $this->assertSame($expectedSQL, str_replace("\n", ' ', $builder->getCompiledSelect()));
    }
```

**Comment by kenjis** [comment]
This PR is very good.
The commit messages are great, with clear explanations!

**Comment by vlakoff** [comment]
I have added this unit test. I made some changes starting from the code you provided:

* put only the "where" condition we are specifically testing
* added "$expectedBinds", for the sake of completeness and to show a bind isn't registered

**Comment by kenjis** [comment]
I tried to think of a use case to test vlakoff@4f702d9, but couldn't come up with one.

**Comment by kenjis** [comment]
Okay, line coverage 100%, Branch coverage 93.75%.

Screenshot 2021-10-30 18 49 39

**Review by kenjis — APPROVED** [review]
I looked at all the commits, and it looks good to me.

File: system/Database/BaseBuilder.php [hunk 1]
```diff
@@ -659,33 +659,34 @@ protected function whereHaving(string $qbKey, $key, $value = null, string $type
                     $op = trim(current($op));
 
                     if (substr($k, -1 * strlen($op)) === $op) {
-                        $k = rtrim(strrev(preg_replace(strrev('/' . $op . '/'), strrev(''), strrev($k), 1)));
+                        $k  = rtrim(substr($k, 0, -1 * strlen($op)));
```
Comment (paulbalandan) [inline system/Database/BaseBuilder.php:662]:
```suggestion
                        $k  = rtrim(substr($k, 0, -strlen($op)));
```

File: system/Database/BaseBuilder.php [hunk 2]
```diff
@@ -659,33 +659,34 @@ protected function whereHaving(string $qbKey, $key, $value = null, string $type
...
+                    $v    = " :{$bind}:";
                 }
             } elseif (! $this->hasOperator($k) && $qbKey !== 'QBHaving') {
                 // value appears not to have been set, assign the test to IS NULL
-                $k .= ' IS NULL';
+                $op = ' IS NULL';
             } elseif (preg_match('/\s*(!?=|<>|IS(?:\s+NOT)?)\s*$/i', $k, $match, PREG_OFFSET_CAPTURE)) {
-                $k = substr($k, 0, $match[0][1]) . ($match[1][0] === '=' ? ' IS NULL' : ' IS NOT NULL');
+                $k  = substr($k, 0, $match[0][1]);
+                $op = ($match[1][0] === '=' ? ' IS NULL' : ' IS NOT NULL');
```
Comment (paulbalandan) [inline system/Database/BaseBuilder.php:683]:
```suggestion
                $op = $match[1][0] === '=' ? ' IS NULL' : ' IS NOT NULL';
```

**Review by paulbalandan — APPROVED** [review]

File: system/Database/BaseBuilder.php [hunk 3]
```diff
@@ -659,7 +659,7 @@ protected function whereHaving(string $qbKey, $key, $value = null, string $type
                     $op = trim(current($op));
 
                     if (substr($k, -1 * strlen($op)) === $op) {
```
Comment (vlakoff) [inline system/Database/BaseBuilder.php:661]: The same change `-strlen` should be applied here as well.

**Comment by vlakoff** [comment]
Force-pushed, integrating the small code style changes from reviews.

**Comment by kenjis** [comment]
@vlakoff Thank you!

**Pull request merged**